File: samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/spots/handlers/DeliveryStatusUpdatedHandler.java
package systems.zlink.samples.deliverydispatch.server.customergateway.spots.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkEntrySpotActorSendHandler;
import systems.zlink.samples.deliverydispatch.server.customergateway.CustomerActor;
import systems.zlink.samples.deliverydispatch.server.customergateway.spots.CustomerEntrySpot;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

public final class DeliveryStatusUpdatedHandler
        implements ZLinkEntrySpotActorSendHandler<
                CustomerEntrySpot, CustomerActor, Messages.DeliveryStatusUpdatedMsg> {
    // --8<-- [start:doc-dd-customer-push]
    @Override
    public CompletionStage<Void> handle(
            CustomerEntrySpot entrySpot,
            CustomerActor actor,
            ZLinkMessageContext context,
            Messages.DeliveryStatusUpdatedMsg message) {
        return actor.push(
                        new Messages.DeliveryStatusNotify(
                                message.deliveryId(),
                                message.status(),
                                message.courierId(),
                                message.occurredAt()))
                .thenRun(
                        () -> {
                            if (message.status() == Messages.DeliveryStatus.Delivered) {
                                System.out.println(
                                        "deliverydispatch-customer pushed status=Delivered"
                                                + " delivery="
                                                + message.deliveryId());
                            }
                        });
    }
    // --8<-- [end:doc-dd-customer-push]
}

File: samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/spots/handlers/DeliveryStatusUpdatedHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.spots.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorSendHandler
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.CustomerActor
import systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.spots.CustomerEntrySpot
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusNotify
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusUpdatedMsg

class DeliveryStatusUpdatedHandler :
    ZLinkSuspendingEntrySpotActorSendHandler<
        CustomerEntrySpot,
        CustomerActor,
        DeliveryStatusUpdatedMsg,
    > {
    // --8<-- [start:doc-dd-customer-push]
    override suspend fun handle(
        entrySpot: CustomerEntrySpot,
        actor: CustomerActor,
        context: ZLinkMessageContext,
        message: DeliveryStatusUpdatedMsg,
    ) {
        actor.push(
            DeliveryStatusNotify(
                deliveryId = message.deliveryId,
                status = message.status,
                courierId = message.courierId,
                occurredAt = message.occurredAt,
            )
        )
        if (message.status.name == "Delivered") {
            println(
                "deliverydispatch-customer pushed status=Delivered " +
                    "delivery=${message.deliveryId}"
            )
        }
    }
    // --8<-- [end:doc-dd-customer-push]
}
